### PR TITLE
fetch-in-react-practice - Kyle Kettler

### PR DESCRIPTION
**Pull request by kyle-kettler**
Kapture 2023-07-18 at 11 06 24

Solution in client-ts

### CODE REVIEW

File: fetch-in-react-practice/client-ts/src/components/Todos.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+};
+
+export default function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function fetchTodos(): Promise<void> {
```
Comment (RobertGardner) [inline fetch-in-react-practice/client-ts/src/components/Todos.tsx:23]: Nice that you are putting on the return types!

File: fetch-in-react-practice/client-ts/src/components/Todos.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todoId: number): Promise<void> {
+    let todoStatus = { isCompleted: false };
+    for (let i = 0; i < todos.length; i++) {
+      if (todoId === todos[i].todoId) {
+        todoStatus.isCompleted = !todos[i].isCompleted
+      }
+    }
```
Comment (RobertGardner) [inline fetch-in-react-practice/client-ts/src/components/Todos.tsx:62]: A lot of JS developers avoid for-loops like this when they can. You could rewrite this using forEach:
```
todos.forEach((todo) => {
  if (todoId === todo.todoId) {
    todoStatus.isCompleted = !todo.isCompleted;
  }
});
```

Another way to do it is to use `find`:
```
todoStatus.isCompleted = !todos.find((t) => t.todoId === todoId)?.isCompleted;
```

File: fetch-in-react-practice/client-ts/src/components/Todos.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      },
+      body: JSON.stringify(todoStatus)
+    });
+    const updatedTodo = await res.json();
+    const updatedTodos = todos.map((item => {
+      if (item.todoId === updatedTodo.todoId) {
+        return updatedTodo;
+      }
+      return item;
+    }));
```
[the remark this answers is not in the document]
Comment (kyle-kettler) [reply fetch-in-react-practice/client-ts/src/components/Todos.tsx:76]: That is a much nicer way to write it. I'll change that and remember it for the future.

File: fetch-in-react-practice/client-ts/src/components/Todos.tsx (hunk 2 above)
Comment (kyle-kettler) [reply fetch-in-react-practice/client-ts/src/components/Todos.tsx:62]: These options are a lot less typing, I'll remember these too! I need to read up on `find` to fully understand how that one works.

**Pull request merged**